src/util/cast.js

```diff
@@ -125,4 +125,48 @@ Cast.isInt = function (val) {
     return false;
 };
 
+Cast.LIST_INVALID = 'INVALID';
+Cast.LIST_ALL = 'ALL';
+/**
+ * Compute a 1-based index into a list, based on a Scratch argument.
+ * Two special cases may be returned:
+ * LIST_ALL: if the block is referring to all of the items in the list.
+ * LIST_INVALID: if the index was invalid in any way.
+ * @param {*} index Scratch arg, including 1-based numbers or special cases.
+ * @param {number} length Length of the list.
+ * @param {boolean} useRound If set, Math.round (not Math.floor for 2.0 compat).
+ * @return {(number|string)} 1-based index for list, LIST_ALL, or LIST_INVALID.
+ */
+Cast.toListIndex = function (
+    index, length, useRound) {
+    if (typeof index !== 'number') {
+        if (index == 'all') {
+            return Cast.LIST_ALL;
+        }
+        if (index == 'last') {
+            if (length > 0) {
+                return length;
+            }
+            return Cast.LIST_INVALID;
+        } else if (index == 'random' || index == 'any') {
+            if (length > 0) {
+                return 1 + Math.floor(Math.random() * length);
+            }
+            return Cast.LIST_INVALID;
+        }
+        if (isNaN(Number(index))) {
```

Comment:
Yep, I think I've convinced myself. I think the isNaN(Number()) test is actually necessary in 2.0 because the value is never coerced to a number there (i.e., no equivalent of `Cast.toNumber` ever happens). Because we now do `Cast.toNumber` below, and that will force any non-number-like values to 0, which then gets returned as Cast.LIST_INVALID... I think it ends up being equivalent. I'll take it out.